smart_open/s3.py

```diff
@@ -419,12 +426,29 @@ def read(self, size=-1):
         if self._position >= self._content_length:
             return b''
 
-        try:
-            binary = self._read_from_body(size)
-        except botocore.exceptions.IncompleteReadError:
-            # The underlying connection of the self._body was closed by the remote peer.
-            self._open_body()
-            binary = self._read_from_body(size)
+        binary = None
+        for attempt, seconds in enumerate(self._sleep_seconds, 1):
+            try:
+                binary = self._read_from_body(size)
+            except botocore.exceptions.BotoCoreError as error:
+                if attempt == len(self._sleep_seconds):
+                    raise
+
+                #
+                # Encountered a low-level connection error.  Attempt to
+                # recover by sleeping it off and retrying.
+                #
+                logger.error(
+                    'caught %r, retrying %d more times',
```

Comment:
Let's also log the sleep time for the next attempt.

```suggestion
                    'caught %r, will automatically retry in %.2fs (%d more attempts left)',
```